examples/controlnet/train_controlnet_sdxl.py

```diff
@@ -0,0 +1,1246 @@
+        weight_decay=args.adam_weight_decay,
+        eps=args.adam_epsilon,
+    )
+
+    # For mixed precision training we cast the text_encoder and vae weights to half-precision
+    # as these models are only used for inference, keeping weights in full precision is not required.
+    weight_dtype = torch.float32
+    if accelerator.mixed_precision == "fp16":
+        weight_dtype = torch.float16
+    elif accelerator.mixed_precision == "bf16":
+        weight_dtype = torch.bfloat16
+
+    # Move vae, unet and text_encoder to device and cast to weight_dtype
+    # The VAE is in float32 to avoid NaN losses.
+    if args.pretrained_vae_model_name_or_path is not None:
+        vae.to(accelerator.device, dtype=weight_dtype)
+    else:
+        vae.to(accelerator.device, dtype=torch.float32)
+    unet.to(accelerator.device, dtype=weight_dtype)
+    text_encoder_one.to(accelerator.device, dtype=weight_dtype)
+    text_encoder_two.to(accelerator.device, dtype=weight_dtype)
+
+    # Here, we compute not just the text embeddings but also the additional embeddings
+    # needed for the SD XL UNet to operate.
+    def compute_embeddings(batch, proportion_empty_prompts, text_encoders, tokenizers, is_train=True):
+        original_size = (args.resolution, args.resolution)
+        target_size = (args.resolution, args.resolution)
+        crops_coords_top_left = (args.crops_coords_top_left_h, args.crops_coords_top_left_w)
+        prompt_batch = batch[args.caption_column]
+
+        prompt_embeds, pooled_prompt_embeds = encode_prompt(
+            prompt_batch, text_encoders, tokenizers, proportion_empty_prompts, is_train
+        )
+        add_text_embeds = pooled_prompt_embeds
+
+        # Adapted from pipeline.StableDiffusionXLPipeline._get_add_time_ids
+        add_time_ids = list(original_size + crops_coords_top_left + target_size)
+        add_time_ids = torch.tensor([add_time_ids])
+
+        prompt_embeds = prompt_embeds.to(accelerator.device)
+        add_text_embeds = add_text_embeds.to(accelerator.device)
+        add_time_ids = add_time_ids.repeat(len(prompt_batch), 1)
+        add_time_ids = add_time_ids.to(accelerator.device, dtype=prompt_embeds.dtype)
+        unet_added_cond_kwargs = {"text_embeds": add_text_embeds, "time_ids": add_time_ids}
+
+        return {"prompt_embeds": prompt_embeds, **unet_added_cond_kwargs}
+
+    # Let's first compute all the embeddings so that we can free up the text encoders
+    # from memory.
+    text_encoders = [text_encoder_one, text_encoder_two]
+    tokenizers = [tokenizer_one, tokenizer_two]
+    train_dataset = get_train_dataset(args, accelerator)
+    compute_embeddings_fn = functools.partial(
+        compute_embeddings,
+        text_encoders=text_encoders,
+        tokenizers=tokenizers,
+        proportion_empty_prompts=args.proportion_empty_prompts,
+    )
+    with accelerator.main_process_first():
+        train_dataset = train_dataset.map(compute_embeddings_fn, batched=True)
```

Comment:
I think this is taken from the docs here https://huggingface.co/docs/accelerate/concept_guides/deferring_execution#applying-a-multiworker-cpu-operation but I'm not sure how caching results from one dataset.map could actually be propagated cross processes if they're all children of the same parent process? Can we confirm what's going on under the hood here?

Also the docs note that this is for cpu operations but we use the gpu to encode the text prompts. I think we should dig in a bit on this

Reply:
This is actually a good point. In my experimentation, I had the following observation on GPU memory usage, my AWS instance has 4 GPUs with 24G each. 
Before map: cuda0,1,2,3 shows 7G taken for each
At the beginning of the ‘map’: cuda0 increased to 17G, and cuda1,2,3 remains 7G each
Later on: cuda0 decreased to 7G and cuda1,2,3 all increased to 17G each
Later on: all GPUs reached to 24G and the training throws OOM. 

Hope the sharing helps

Reply:
@williamberman what do you want me to confirm? 

Cc: @muellerzr what would be the best course of action here? briefly, I am trying to map a function (that uses a text encoder placed on a GPU) to precompute the embeddings.

Reply:
It's not worth it for now to change this here IMO, so I would leave as is.

What happens here is that we **don't** parallelize over the data but just run `compute_embeddings_fn` over the whole train dataset and then cache the output as a file on hard drive.

 When doing this it makes no sense to do this for every process because it would be unnecessary work and each process would overwrite each other's cache file.

So we have two options here:
1.) We use a Pytorch data loader and correctly parallelize over the data. In this case we should not only run on main process. This would be faster and GPUs would be better utilized, but it's more complex to write and I'm always a bit worried about executing dataset's map function over multiple processes
2.) We leave as is which is what I think we should do here

Reply:
I'm less concerned about correctly parallelizing for efficiency and more concerned about correctness. 

I.e. if datasets is writing to a cache file for a dynamic call to a function and multiple different processes can overwrite the same cache file, that's a bug in datasets regardless of syncronization calls around it (you have no way to tell if the results are going to be the same).  

I'm also concerned that the datasets docs note that the example should be used for if doing cpu pre-processing, if we're doing GPU pre-processing and somehow gpu tensors get shared cross processes, pytorch requires certain conventions to be handled https://pytorch.org/docs/stable/notes/multiprocessing.html 

I'm ok with the merge if we get successful training runs on a single gpu but I think there's a few things we're potentially missing here

Reply:
Hi @williamberman, just to kindly confirm - do we indicate that there might be unknown risks if using multi-GPUs for training?

Reply:
I wonder if this section is made optional, and is it reasonable to cache text in advance when training on a larger dataset? Like laion-5b, which seems a little strange?

Reply:
I think it makes sense to cache the part that is not going to be changed during the training process. 

I am ccing @lhoestq and @muellerzr (Datasets and Accelerate teams respectively) to provide us with some guidance here. 

Quentin, Zach,

Could you share your perspectives on how the embedding computation function is being mapped here considering https://github.com/huggingface/diffusers/pull/4038#discussion_r1267144289?

Reply:
> I.e. if datasets is writing to a cache file for a dynamic call to a function and multiple different processes can overwrite the same cache file, that's a bug in datasets regardless of syncronization calls around it (you have no way to tell if the results are going to be the same).

The first process runs the computation and cache it, and the other processes reload from cache.
So all the processes end up sith the same data.

> I'm also concerned that the datasets docs note that the example should be used for if doing cpu pre-processing, if we're doing GPU pre-processing and somehow gpu tensors get shared cross processes, pytorch requires certain conventions to be handled https://pytorch.org/docs/stable/notes/multiprocessing.html

Here you're doing `map` without multiprocessing so you're fine (no `num_proc=` argment passed to `map`)

Reply:
Minor but I'd recommend passing the fingerprint hash yourself for the cache.
This way it won't have to hash a full model to compute the fingerprint of the resulting dataset
```python
    with accelerator.main_process_first():
        from datasets.fingerprint import Hasher
        new_fingerprint = Hasher.hash(args)  # fingerprint used by the cache for the other processes to load the result
        train_dataset = train_dataset.map(compute_embeddings_fn, batched=True, new_fingerprint=new_fingerprint))
```

Reply:
@lhoestq perfect exactly what I was looking for :) If you don't mind me asking, if we don't pass the fingerprint for the cache into .map, how does it determine what fingerprint to use? It seems a bit odd that we could call .map from multiple processes and they would end up with the same fingerprint. I.e. we have no way to determine that the function to .map creates the same results in the different processes

Reply:
@sayakpaul could you do a follow up PR and manually pass the fingerprint to .map? That would fix all my concerns

Reply:
> If you don't mind me asking, if we don't pass the fingerprint for the cache into .map, how does it determine what fingerprint to use? It seems a bit odd that we could call .map from multiple processes and they would end up with the same fingerprint. I.e. we have no way to determine that the function to .map creates the same results in the different processes

We use `dill` to pickle and hash the function that you pass to `map` and get a deterministic hash that doesn't depend on the process

Reply:
@lhoestq Are all parameters necessary to calculate fingerprint? A slight change will cause the cache to be recalculated, which is too painful for large data sets.

Reply:
Ideally you should only hash the parameters that influence the preprocessing indeed